src/components/form/HdForm.vue

```diff
@@ -0,0 +1,90 @@
+<template>
+  <form class="form" @submit.prevent="submit" novalidate>
+    <slot />
+  </form>
+</template>
+
+<script>
+import _cloneDeep from 'lodash/cloneDeep';
+import _isEqual from 'lodash/isEqual';
+import _isNil from 'lodash/isNil';
+import { scrollToEl } from 'homeday-blocks/src/services/scrolling';
+import { formatNestedData } from 'homeday-blocks/src/services/utils';
+
+export default {
+  name: 'HdForm',
+  provide() {
+    return {
+      addFormField: this.addField,
+      removeFormField: this.removeField,
+    };
+  },
+  props: {
+    scrollToInvalidField: {
+      type: Boolean,
+      default: false,
+    },
+  },
+  data() {
+    return {
+      initialFormData: {},
+      fields: [],
+    };
+  },
+  created() {
+    this.initialFormData = _cloneDeep(this.getFormData());
+  },
+  methods: {
+    getFormData() {
+      return this.fields.reduce((formData, field) => ({
+        ...formData,
+        [field.name]: field.value,
+      }), {});
+    },
+    getInvalidFields() {
+      return this.fields.filter((field) => {
+        if (field.validate) {
+          return !field.validate();
+        }
+        return false;
+      });
+    },
+    validate() {
+      return this.getInvalidFields().length === 0;
+    },
+    submit() {
+      const invalidFields = this.getInvalidFields();
+      const isValid = invalidFields.length === 0;
+      const formData = formatNestedData(this.getFormData());
+      const hasChanged = !_isEqual(formData, this.initialFormData);
+
+      this.$emit('submit', {
+        formData,
+        isValid,
+        hasChanged,
+        invalidFields,
+      });
+
+      if (!isValid && this.scrollToInvalidField) {
+        scrollToEl({ el: invalidFields[0].$el });
+      }
+    },
+    addField(fieldToAdd) {
+      if (_isNil(fieldToAdd.name)) {
+        console.warn('The field must have a `name` property to be added to the HdForm', fieldToAdd);
+        return;
+      }
+
+      this.fields.push(fieldToAdd);
+
+      // If a new field was dynamically added to the form, we add its value to the initial form data
+      if (this.initialFormData[fieldToAdd.name] === undefined) {
+        this.initialFormData[fieldToAdd.name] = _cloneDeep(fieldToAdd.value);
+      }
+    },
```

Comment:
I'm not a huge fan of comments. I tend to see them as code smells. I love [this reference](https://refactoring.guru/smells/comments) in particular, I love this statement:

> The best comment is a good name for a method or class.

And, on the `addField` method, I think that this statement applies:

> If a comment explains a section of code, this section can be turned into a separate method via Extract Method. The name of the new method can be taken from the comment text itself, most likely.

So I suggest the following change:

```suggestion
    addField(fieldToAdd) {
      if (_isNil(fieldToAdd.name)) {
        console.warn('The field must have a `name` property to be added to the HdForm', fieldToAdd);
        return;
      }

      this.fields.push(fieldToAdd);
      
      if (this.isDynamicallyAddedField(fieldToAdd.name)) {
        this.initialFormData[fieldToAdd.name] = _cloneDeep(fieldToAdd.value);
      }
    },
    isDynamicallyAddedField(fieldName) {
      return this.initialFormData[fieldName] === undefined;
    },
```

Of course, I don't mean that all comments should be banned. I love [this article](https://css-tricks.com/the-art-of-comments/) by Sarah Drasner where she writes about different kind of comments, and she wrote:

> Code can describe how, but it cannot explain why.

Reply:
I like the approach 👍